### PR TITLE
2068 support drain delegate operation in contract api

### PR DESCRIPTION
**Pull request by hui-an-yang**
Thank you for your contribution to Taquito.

Before submitting this PR, please make sure:

- [ ] Your code builds cleanly without any errors or warnings
- [ ] You have added unit tests
- [ ] You have added integration tests (if relevant/appropriate)
- [ ] All public methods or types have TypeDoc coverage with a complete description, and ideally an @example
- [ ] You have added or updated corresponding documentation
- [ ] If relevant, you have written a first draft summary describing the change for inclusion in Release Notes. 

## Release Note Draft Snippet

__If relevant, please write a summary of your change that will be suitable for
inclusion in the Release Notes for the next Taquito release.__

### CODE REVIEW

File: packages/taquito/src/contract/rpc-contract-provider.ts [hunk 1]
```diff
@@ -60,7 +62,7 @@ import { TxRollupOriginationOperation } from '../operations/tx-rollup-originatio
 import { TxRollupBatchOperation } from '../operations/tx-rollup-batch-operation';
 import { TransferTicketOperation } from '../operations/transfer-ticket-operation';
 import { IncreasePaidStorageOperation } from '../operations/increase-paid-storage-operation';
-import { BallotOperation } from '../operations';
+import { BallotOperation, DrainDelegateOperation } from '../operations';
```
Comment (dsawali) [inline packages/taquito/src/contract/rpc-contract-provider.ts:65]: I think we should separate the imports like the others. I think I've made a mistake on the BallotOperation import path. It should be its own line for clarity.

File: integration-tests/sandbox-drain-delegate-operation.spec.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    })
+    flextesanet('Should be able to inject drain_delegate operation', async (done) => {
+      const drainOp = await Destination.contract.drainDelegate({
+        consensus_key: destinationPkh,
+        delegate: delegatePkh,
+        destination: destinationPkh,
+      });
+      await drainOp.confirmation();
+      
+      expect((await Delegate.rpc.getBalance(delegatePkh)).toNumber()).toEqual(0);
```
Comment (dsawali) [inline integration-tests/sandbox-drain-delegate-operation.spec.ts:51]: could you add more assertions to check whether the op has been injected in the block?

i.e. `includedInBlock`, `status` or something similar?

File: integration-tests/sandbox-drain-delegate-operation.spec.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      expect((await Destination.rpc.getBalance(destinationPkh)).toNumber()).toBeGreaterThan(0);
```
Comment (dsawali) [inline integration-tests/sandbox-drain-delegate-operation.spec.ts:52]: if balance of the previous baker was `x` and the balance of the destination was `y`, we should check whether the destinationPkh balance is now `x+y`

File: integration-tests/sandbox-drain-delegate-operation.spec.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
+import { TezosToolkit } from "@taquito/taquito";
+import { CONFIGS, sleep } from "./config";
+
+CONFIGS().forEach(({ lib, rpc, setup, createAddress }) => {
+  const Tezos = lib;
+  const flextesanet = (rpc === 'http://localhost:20001') ? it : it.skip;
```
Comment (dsawali) [inline integration-tests/sandbox-drain-delegate-operation.spec.ts:6]: we could maybe specify which protocol it is by renaming this to `flextesaLimanet`

**Review by dsawali — CHANGES_REQUESTED** [review]

File: integration-tests/sandbox-env-lima.sh [hunk 5]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+export RUN_LIMANET_WITH_SECRET_KEY=true
```
[the remark this answers is not in the document]
Comment (hui-an-yang) [reply integration-tests/sandbox-env-lima.sh:2]: Didn't know it's the common understanding, will change it back

File: packages/taquito/src/contract/rpc-contract-provider.ts (hunk 1 above)
Comment (hui-an-yang) [reply packages/taquito/src/contract/rpc-contract-provider.ts:65]: Will update both, was wondering do we export them in `operations/index.ts` only for end users not for code base?

File: packages/taquito/src/contract/interface.ts [hunk 6]
```diff
@@ -221,6 +223,16 @@ export interface ContractProvider extends StorageProvider {
    */
   txRollupSubmitBatch(params: TxRollupBatchParams): Promise<TxRollupBatchOperation>;
 
+  /**
+   *
+   * @description Submit drain delegate for an ongoing proposal
```
[the remark this answers is not in the document]
Comment (hui-an-yang) [reply packages/taquito/src/contract/interface.ts:228]: sorry, it'll be updated

File: integration-tests/sandbox-drain-delegate-operation.spec.ts (hunk 4 above)
Comment (hui-an-yang) [reply integration-tests/sandbox-drain-delegate-operation.spec.ts:6]: thanks, updated to flextesaLimaNet

File: integration-tests/sandbox-drain-delegate-operation.spec.ts (hunk 2 above)
Comment (hui-an-yang) [reply integration-tests/sandbox-drain-delegate-operation.spec.ts:51]: will add includedInBlock to beDefined

**Comment by dsawali** [comment]
I think we should leverage Dan's PR to use multiple bootstrap accounts. That way we remove the risk of failures in other tests due to the drain delegate op. PR : https://github.com/ecadlabs/taquito/pull/2219

File: packages/taquito/src/contract/interface.ts [hunk 7]
```diff
@@ -221,6 +223,16 @@ export interface ContractProvider extends StorageProvider {
    */
   txRollupSubmitBatch(params: TxRollupBatchParams): Promise<TxRollupBatchOperation>;
 
+  /**
+   *
+   * @description Submit drain delegate operation
```
Comment (zainen) [inline packages/taquito/src/contract/interface.ts:228]: Submits a drain delegate operation i think is a bit more proper

File: integration-tests/sandbox-drain-delegate-operation.spec.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
+import { TezosToolkit } from "@taquito/taquito";
+import { CONFIGS, sleep } from "./config";
+
+CONFIGS().forEach(({ lib, rpc, setup, createAddress }) => {
+  const Tezos = lib;
+  const flextesaLimaNet = (rpc === 'http://localhost:20000') ? it : it.skip;
```
[the remark this answers is not in the document]
Comment (hui-an-yang) [reply integration-tests/sandbox-drain-delegate-operation.spec.ts:6]: Think you have a point, can see how `flextesanet` will be less coding debt when we adapt into Lima

File: packages/taquito/src/contract/interface.ts (hunk 7 above)
Comment (hui-an-yang) [reply packages/taquito/src/contract/interface.ts:228]: thanks, will push a new commit

File: docs/consensus_key.md [hunk 9]
```diff
@@ -20,3 +20,20 @@ await op.confirmation();
 ```
 - `pk` is the public key you want the consensus key to point to
 
+
+## Drain Delegate
+This is an operation that must be signed by the active consensus key `consensus_pkh` of the baker `baker_pkh`. This operation immediately transfers all the spendable balance of the `baker_pkh`’s implicit account into the `destination_pkh` implicit account. It has no effect on the frozen balance. This operation is included in pass 2 (anonymous operations). So drain operations don’t compete with regular manager operations for gas and block size quota; the 1M restriction (one-operation-per-manager-per-block) applies to drain operations as well, meaning that a drain for a baker and a transfer operation from the same baker are in conflict. As an incentive for bakers to include drain operations, a fixed fraction of the drained baker’s spendable balance is transferred as fees to the baker that includes the operation, i.e. the maximum between 1tz or 1% of the spendable balance.
```
[the remark this answers is not in the document]
Comment (hui-an-yang) [reply docs/consensus_key.md:25]: Thank you, with this doc line 6 Davis have left reference to general consensus_key release note, is it still better to add reference for drain_delegate op part?

File: packages/taquito/src/operations/drain-delegate-operation.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    private readonly params: RPCDrainDelegateOperation,
+    raw: ForgedBytes,
+    results: OperationContentsAndResult[],
+    context: Context
+  ) {
+    super(hash, raw, results, context);
+  }
+
+  get operationResults() {
+    const DrainDelegateOp =
```
[the remark this answers is not in the document]
Comment (hui-an-yang) [reply packages/taquito/src/operations/drain-delegate-operation.ts]: pushed a commit to address this

**Review by roxaneletourneau — APPROVED** [review]
Lgtm  :)

**Review by dsawali — APPROVED** [review]

**Pull request merged**